File: 2024/day3/src/main.py
import re


def main():
    with open("input.txt") as f:
        data = f.read()

    pattern = r"mul\s*\(\s*(-?\d+)\s*,\s*(-?\d+)\s*\)"
    active = True
    total = 0

    split_pattern = r"(do\(\)|don\'t\(\))"
    sections = re.split(split_pattern, data)

    for section in sections:
        if section == "do()":
            active = True
        elif section == "don't()":
            active = False
        else:
            if active:
                matches = re.findall(pattern, section)
                for a, b in matches:
                    product = int(a) * int(b)
                    total += product
    print(total)


if __name__ == "__main__":
    main()

File: 2024/day3/src/main.rs
fn main() {
    let input = std::fs::read_to_string("input.txt").unwrap();

    let section_re = regex::Regex::new(r#"(do\(\)|don't\(\))"#).unwrap();
    let sections = section_re.split(&input);

    let mut active = true;
    let mut total = 0;

    let mul_re = regex::Regex::new(r#"mul\s*\(\s*(-?\d+)\s*,\s*(-?\d+)\s*\)"#)
        .expect("Invalid multiplication regex");

    for section in sections {
        let section = section.trim();

        if section == "do()" {
            active = true;
        } else if section == "don't()" {
            active = false;
        } else {
            if active {
                for nums in mul_re.captures_iter(section) {
                    total += nums[1].parse::<i32>().unwrap() * nums[2].parse::<i32>().unwrap();
                }
            }
        }
    }

    println!("{}", total);
}
